native/src/main/java/io/ballerina/lib/data/csvdata/utils/CsvUtils.java

```diff
@@ -0,0 +1,373 @@
+                 TypeTags.UNSIGNED16_INT_TAG, TypeTags.UNSIGNED32_INT_TAG -> true;
+            default -> false;
+        };
+    }
+
+    public static String[] createHeadersForParseLists(BArray csvElement, String[] headers, CsvConfig config) {
+        Object customHeaders = config.customHeaders;
+        long headerRows = config.headerRows;
+
+        int length = headers.length;
+        if (customHeaders instanceof BArray array) {
+            if (array.size() != length) {
+                throw DiagnosticLog.error(DiagnosticErrorCode.INVALID_CUSTOM_HEADER_LENGTH);
+            }
+            for (int i = 0; i < length; i++) {
+                headers[i] = array.get(i).toString();
+            }
+            return headers;
+        }
+
+        if (headerRows == 1) {
+            return csvElement.getStringArray();
+        }
+
+        if (headerRows > 1) {
+            throw DiagnosticLog.error(DiagnosticErrorCode.NO_CUSTOM_HEADER_PROVIDED);
+        }
+
+        // when headerRows = 0 and customHeaders = null
+        for (int i = 0; i < length; i++) {
+            headers[i] = String.valueOf(i + 1);
+        }
+        return headers;
+    }
+
+    public static Optional<Type> getMutableType(IntersectionType intersectionType) {
+        for (Type constituentType : intersectionType.getConstituentTypes()) {
+            if (constituentType.getTag() == TypeTags.READONLY_TAG) {
+                continue;
+            }
+            return Optional.of(constituentType);
+        }
+        return Optional.empty();
+    }
+
+    public static Object convertToBasicType(Object csv, Type targetType, CsvConfig config) {
+        if (csv == null) {
+            csv = config.nilValue;
+        }
+        return ValueUtils.convert(csv, targetType);
+    }
+
+    public static void checkRequiredFieldsAndLogError(Map<String, Field> filedHierarchy, boolean absentAsNilableType) {
+        filedHierarchy.values().forEach(field -> {
+            if (absentAsNilableType && field.getFieldType().isNilable()) {
+                return;
+            }
+            if (SymbolFlags.isFlagOn(field.getFlags(), SymbolFlags.REQUIRED)) {
+                throw DiagnosticLog.error(DiagnosticErrorCode.INVALID_FIELD_IN_CSV, field.getFieldName());
+            }
+        });
+    }
+
+    public static boolean isHeaderFieldsEmpty(Map<String, Field> currentField) {
+        for (Field field: currentField.values()) {
+            if (SymbolFlags.isFlagOn(field.getFlags(), SymbolFlags.REQUIRED)) {
+                return false;
+            }
+        }
+        return true;
+    }
+
+    public static boolean checkTypeCompatibility(Type constraintType, Object csv, boolean stringConversion) {
+        int tag = constraintType.getTag();
+        if (csv instanceof BString) {
+            if (stringConversion || TypeTags.isStringTypeTag(tag) || isJsonOrAnyDataOrAny(tag)) {
+                return true;
+            }
```

Comment:
I think dirctly returning the condition here is just as readable here, WDYT?
```suggestion
        return stringConversion || TypeTags.isStringTypeTag(tag) || isJsonOrAnyDataOrAny(tag);
```

Reply:
Same for subsequent conditions in this method as well